### PR TITLE
ref(cardinality): Log Sentry errors for cardinality limits

### PR DESCRIPTION
**Pull request by Dav1dde**
To also track which namespace/limit was affected, this PR also adds a hashset of limit IDs to the cardinality limiter return value.

#skip-changelog

### CODE REVIEW

File: relay-config/src/config.rs [hunk 1]
```diff
@@ -1291,12 +1291,21 @@ pub struct CardinalityLimiter {
...
     /// Defaults to 180 seconds, 3 minutes.
     pub cache_vacuum_interval: u64,
+
+    /// Sample rate for Cardinality Limiter Sentry errors.
+    ///
+    /// Rate needs to be between `0.0` and `1.0`.
+    /// If set to `1.0` all cardinality limiter rejections will be logged as a Sentry error.
+    ///
+    /// Defaults to 1% (0.01).
+    pub error_sample_rate: f32,
```
Comment (jjbayer) [inline relay-config/src/config.rs]: Not a blocker, but should we put this into global config instead? Would be nice to control it via automator.

**Review by jjbayer — APPROVED** [review]

File: relay-server/src/services/processor.rs [hunk 2]
```diff
@@ -2251,6 +2263,13 @@ impl UpstreamRequest for SendEnvelope {
     }
 }
 
+/// Returns a boolean wether the current item should be sampled or discarded
+/// using the passed `rate` (0 <= rate <= 1).
```
Comment (iker-barriocanal) [inline relay-server/src/services/processor.rs]:
```suggestion
/// Returns a boolean whether the current item should be sampled or discarded
/// using the passed `rate` (0 <= rate <= 1).
```

**Review by iker-barriocanal — APPROVED** [review]

File: relay-dynamic-config/src/global.rs [hunk 3]
```diff
@@ -79,6 +74,16 @@ pub struct Options {
...
 
+    /// Sample rate for Cardinality Limiter Sentry errors.
+    ///
+    /// Rate needs to be between `0.0` and `1.0`.
+    /// If set to `1.0` all cardinality limiter rejections will be logged as a Sentry error.
+    #[serde(
+        rename = "relay.cardinality-limiter.error-sample-rate",
+        skip_serializing_if = "is_default"
+    )]
+    pub cardinality_limiter_error_sample_rate: f32,
```
Comment (jjbayer) [inline relay-dynamic-config/src/global.rs:85]: We could theoretically make this `#[cfg(processing)]` (probably other cardinality limiter options as well).

File: relay-server/src/services/processor.rs [hunk 4]
```diff
@@ -2251,6 +2262,13 @@ impl UpstreamRequest for SendEnvelope {
     }
 }
 
+/// Returns a boolean whether the current item should be sampled or discarded
```
Comment (jjbayer) [inline relay-server/src/services/processor.rs]: nit: Didn't find a reference for this but this seems to be the Rust way.
```suggestion
/// Returns `true` if the current item should be sampled.
```

File: relay-cardinality/src/limiter.rs [hunk 5]
```diff
@@ -136,30 +136,46 @@ impl<T: Limiter> CardinalityLimiter<T> {
...
+    fn new(source: Vec<T>, rejections: RejectionTracker<'a>) -> Self {
         Self {
             source,
-            rejections: rejections.0,
+            rejections: rejections.entries,
+            limits: rejections.limits,
         }
     }
 
+    /// Whether any items are rejected.
```
Comment (jjbayer) [inline relay-cardinality/src/limiter.rs]:
```suggestion
    /// Returns `true` if any items have been rejected.
```

**Review by jjbayer — APPROVED** [review]

File: relay-dynamic-config/src/global.rs (hunk 3 above)
Comment (Dav1dde) [reply relay-dynamic-config/src/global.rs:85]: Not a fan of this, it would just end up being collected in the fallback hashmap instead

**Pull request merged**